akka-http-core/src/main/scala/akka/http/impl/engine/http2/Http2Demux.scala

```diff
@@ -211,22 +211,45 @@ private[http2] abstract class Http2Demux(http2Settings: Http2CommonSettings, ini
   def wrapTrailingHeaders(headers: ParsedHeadersFrame): Option[HttpEntity.ChunkStreamPart]
   def completionTimeout: FiniteDuration
 
-  def createLogic(inheritedAttributes: Attributes): GraphStageLogic =
-    new TimerGraphStageLogic(shape) with Http2MultiplexerSupport with Http2StreamHandling with GenericOutletSupport with StageLogging with LogHelper {
+  override def createLogicAndMaterializedValue(inheritedAttributes: Attributes): (GraphStageLogic, ServerTerminator) = {
+    object Logic extends TimerGraphStageLogic(shape) with Http2MultiplexerSupport with Http2StreamHandling with GenericOutletSupport with StageLogging with LogHelper with ServerTerminator {
       logic =>
 
       import Http2Demux.CompletionTimeout
 
       def wrapTrailingHeaders(headers: ParsedHeadersFrame): Option[HttpEntity.ChunkStreamPart] = stage.wrapTrailingHeaders(headers)
 
       override def isServer: Boolean = stage.isServer
+
       override def settings: Http2CommonSettings = http2Settings
+
       override def isUpgraded: Boolean = upgraded
 
       override protected def logSource: Class[_] = if (isServer) classOf[Http2ServerDemux] else classOf[Http2ClientDemux]
+
       // cache debug state at the beginning to avoid that this has to be queried all the time
       override lazy val isDebugEnabled: Boolean = super.isDebugEnabled
 
+      private val terminationPromise = Promise[Http.HttpTerminated]()
+      private var terminating: Boolean = false
+      private var lastIdBeforeTermination: Int = 0
+      private val terminateCallback = getAsyncCallback[FiniteDuration](triggerTermination)
+      override def terminate(deadline: FiniteDuration)(implicit ex: ExecutionContext): Future[Http.HttpTerminated] = {
+        terminateCallback.invoke(deadline)
+        terminationPromise.future
+      }
+      private def triggerTermination(deadline: FiniteDuration): Unit = {
+        // check if we are already terminating, otherwise start termination
+        if (!terminating) {
+          log.debug("Termination now starting...")
+          terminating = true
+          pushGOAWAY(ErrorCode.NO_ERROR, "")
```

Comment:
I wonder if we should introduce a `debug` parameter to `triggerTermination`, might be helpful in debugging later... but perhaps not important for now.

Reply:
This seems like an implementation detail that is not really relevant for the other side, is it? After all, what we do here is close the connection because we want to which is a completely valid thing to do...

Reply:
I now called it a "Voluntary connection close.". Whatever that will mean for the other side...